src/libstd/sync/remutex.rs

```diff
@@ -0,0 +1,248 @@
+    /// acquired.
+    pub fn lock(&self) -> LockResult<ReentrantMutexGuard<T>> {
+        unsafe { self.inner.lock() }
+        ReentrantMutexGuard::new(&self)
+    }
+
+    /// Attempts to acquire this lock.
+    ///
+    /// If the lock could not be acquired at this time, then `Err` is returned.
+    /// Otherwise, an RAII guard is returned.
+    ///
+    /// This function does not block.
+    ///
+    /// # Failure
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return failure if the mutex would otherwise be
+    /// acquired.
+    pub fn try_lock(&self) -> TryLockResult<ReentrantMutexGuard<T>> {
+        if unsafe { self.inner.try_lock() } {
+            Ok(try!(ReentrantMutexGuard::new(&self)))
+        } else {
+            Err(TryLockError::WouldBlock)
+        }
+    }
+
+    /// Determine whether the lock is poisoned.
+    ///
+    /// If another thread is active, the lock can still become poisoned at any
+    /// time.  You should not trust a `false` value for program correctness
+    /// without additional synchronization.
+    #[inline]
+    pub fn is_poisoned(&self) -> bool {
+        self.poison.get()
+    }
+}
+
+#[unsafe_destructor]
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+impl<T> Drop for ReentrantMutex<T> {
+    fn drop(&mut self) {
+        // This is actually safe b/c we know that there is no further usage of
+        // this mutex (it's up to the user to arrange for a mutex to get
+        // dropped, that's not our job)
+        unsafe { self.inner.destroy() }
+    }
+}
+
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+impl<T: fmt::Debug + 'static> fmt::Debug for ReentrantMutex<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.try_lock() {
+            Ok(guard) => write!(f, "ReentrantMutex {{ data: {:?} }}", *guard),
+            Err(TryLockError::Poisoned(err)) => {
+                write!(f, "ReentrantMutex {{ data: Poisoned({:?}) }}", **err.get_ref())
+            },
+            Err(TryLockError::WouldBlock) => write!(f, "ReentrantMutex {{ <locked> }}")
+        }
+    }
+}
+
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+impl<'mutex, T> ReentrantMutexGuard<'mutex, T> {
+    fn new(lock: &'mutex ReentrantMutex<T>)
+            -> LockResult<ReentrantMutexGuard<'mutex, T>> {
+        poison::map_result(lock.poison.borrow(), |guard| {
+            ReentrantMutexGuard {
+                __lock: lock,
+                __poison: guard,
+            }
+        })
+    }
+}
+
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+impl<'mutex, T> Deref for ReentrantMutexGuard<'mutex, T> {
+    type Target = T;
+
+    fn deref<'a>(&'a self) -> &'a T {
+        unsafe { &*self.__lock.data.get() }
+    }
+}
+
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+impl<'mutex, T> DerefMut for ReentrantMutexGuard<'mutex, T> {
```

Comment:
Uh, this hurts a lot, since mutable references are exactly what most stdio methods want and alternatives I’ve considered so far:

* with_mut_ref(&self, fucn: fn(&mut T) -> R) -> R;
* as_mut_ptr(&self) -> *mut T;

are both insufficient. First one prevents aliasing reliably by disallowing to enclose over the environment, at the same time making the whole scheme as useless as it is safe; and the second one allows keeping pointer to the data after the lock expires.

I fear it might be not viable to have `ReentrantMutex` which protects data inside itself…